geomstats/discretized_curves_space.py

```diff
@@ -0,0 +1,424 @@
+            for k in range(len(t)):
+                point_at_time_t[k, :] = self.exp(
+                        tangent_vec=tangent_vecs[k, :],
+                        base_point=new_initial_point)
+            return point_at_time_t
+
+        return point_on_geodesic
+
+
+class SRVMetric(RiemannianMetric):
+    """
+    Elastic metric defined using the Square Root Velocity Function
+    (see Srivastava et al. 2011).
+    """
+    def __init__(self, embedding_manifold):
+        super(SRVMetric, self).__init__(
+                dimension=embedding_manifold.dimension,
+                signature=(embedding_manifold.dimension, 0, 0))
+        self.embedding_metric = embedding_manifold.metric
+        self.L2_metric = L2Metric(embedding_manifold=embedding_manifold)
+
+    def pointwise_inner_product(self, tangent_vec_a, tangent_vec_b,
+                                base_point):
+        """
+        Compute the inner products of the components of a (series of)
+        pair(s) of tangent vectors at (a) base point(s).
+        """
+        point_ndim = 2
+        n_sampling_points = base_point.shape[-2]
+        base_point = gs.to_ndarray(base_point, to_ndim=point_ndim + 1)
+        tangent_vec_a = gs.to_ndarray(tangent_vec_a, to_ndim=point_ndim + 1)
+        tangent_vec_b = gs.to_ndarray(tangent_vec_b, to_ndim=point_ndim + 1)
+
+        n_tangent_vecs = tangent_vec_a.shape[0]
+        inner_prod = gs.zeros([n_tangent_vecs, n_sampling_points])
+        for k in range(n_tangent_vecs):
+            inner_prod[k, :] = self.embedding_metric.inner_product(
+                    tangent_vec_a[k, :], tangent_vec_b[k, :],
+                    base_point[k, :]).squeeze()
+
+        return inner_prod
+
+    def pointwise_norm(self, tangent_vec, base_point):
+        """
+        Compute the norms of the components of a (series of)
+        tangent vector(s) at (a) base point(s).
+        """
+        sq_norm = self.pointwise_inner_product(tangent_vec_a=tangent_vec,
+                                               tangent_vec_b=tangent_vec,
+                                               base_point=base_point)
+        return gs.sqrt(sq_norm)
+
+    def square_root_velocity(self, curve):
+        """
+        Compute the square root velocity representation of a curve.
+        """
+        n_coords = curve.shape[-1]
+        n_sampling_points = curve.shape[-2]
+        n_points_in_total = np.cumprod(curve.shape)[-2]
+        srv_shape = gs.array(curve.shape)
+        srv_shape[-2] -= 1
+
+        curve = curve.reshape(n_points_in_total, n_coords)
+        velocity = n_sampling_points * self.embedding_metric.log(
+                point=curve[1:, :], base_point=curve[:-1, :])
+        velocity_norm = self.embedding_metric.norm(velocity, curve[:-1, :])
+        assert gs.all(velocity_norm != 0)
+        srv = velocity / gs.sqrt(velocity_norm)
+        index = gs.arange(n_points_in_total - 1)
+        index_select = index[(index + 1) % n_sampling_points != 0]
```

Comment:
what is this index selection?

Reply:
This is again linked to vectorization. When ```curve``` has shape ```[n_curves, n_sampling_points, n_coords]```, it has to be reshaped into ```[n_curves * n_sampling_points, n_coords]``` before computing the velocity using
```
velocity = n_sampling_points * self.embedding_metric.log(
                point=curve[1:, :], base_point=curve[:-1, :])
```

At regular intervals of length ```n_sampling_points```, this computes a log that is meaningless, namely the log between the end point of ```curve[k, :, :]``` and the starting point of the next curve ```curve[k + 1, :, :]```. This index selection gets rid of these vectors.

Reply:
I see, thanks for explaining! Could you add a comment to explain (very rapidly) this for future collaborators? Thanks!